pkg/server/distributor_test.go

```diff
@@ -0,0 +1,475 @@
+	for server, stopErr := range stopErrs {
+		t.Fatalf("unexpected stop error from module server %s: %v", server, stopErr)
+	}
+}
+
+func startAndWaitHealthy(t *testing.T, testServer testModuleServer, runErrs map[string]error) {
+	go func() {
+		// this next line is to avoid double registration, as both InitializeDocumentBuilders as well as ProvideUnifiedStorageGrpcService
+		// are hard-coded to use prometheus.DefaultRegisterer
+		// the alternative would be to get the registry from wire, in which case the tests would receive a new
+		// registry automatically, but that _may_ change metric names
+		// We can remove this once that's fixed
+		prometheus.DefaultRegisterer = prometheus.NewRegistry()
+		if err := testServer.server.Run(); err != nil && !errors.Is(err, context.Canceled) {
+			runErrs[testServer.id] = err
+		}
+	}()
+
+	deadline := time.Now().Add(20 * time.Second)
+	for {
+		require.NoError(t, runErrs[testServer.id], "failed to start "+testServer.id)
+
+		conn, err := net.DialTimeout("tcp", testServer.grpcAddress, 1*time.Second)
+		if err == nil {
+			_ = conn.Close()
+			break
+		}
+
+		if time.Now().After(deadline) {
+			t.Fatal("server failed to become ready: ", testServer.id)
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+
+	res, err := testServer.healthClient.Check(context.Background(), &grpc_health_v1.HealthCheckRequest{})
+	require.NoError(t, err)
+	require.Equal(t, res.Status, grpc_health_v1.HealthCheckResponse_SERVING)
+}
+
+type testModuleServer struct {
+	server         *ModuleServer
+	healthClient   grpc_health_v1.HealthClient
+	resourceClient resource.ResourceClient
+	id             string
+	grpcAddress    string
+}
+
+func initDistributorServerForTest(t *testing.T) testModuleServer {
+	cfg := setting.NewCfg()
+	cfg.HTTPPort = "13000"
+	cfg.GRPCServer.Network = "tcp"
+	cfg.GRPCServer.Address = "127.0.0.1:20000"
+	cfg.EnableSharding = true
+	cfg.MemberlistBindAddr = "127.0.0.1"
+	cfg.MemberlistJoinMember = "127.0.0.1:17946"
+	cfg.MemberlistAdvertiseAddr = "127.0.0.1"
+	cfg.MemberlistAdvertisePort = 17946
+	cfg.Target = []string{modules.Distributor}
+	cfg.InstanceID = "distributor" // does nothing for the distributor but may be useful to debug tests
+
+	conn, err := grpc.NewClient(cfg.GRPCServer.Address,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	require.NoError(t, err)
+	client := resource.NewLegacyResourceClient(conn)
+
+	server := initModuleServerForTest(t, cfg, Options{}, api.ServerOptions{})
+
+	server.resourceClient = client
+
+	return server
+}
+
+func createStorageServerApi(t *testing.T, instanceId int, dbType, dbConnStr string) testModuleServer {
+	cfg := setting.NewCfg()
+	section, err := cfg.Raw.NewSection("database")
+	require.NoError(t, err)
+
+	_, err = section.NewKey("type", dbType)
+	require.NoError(t, err)
+	_, err = section.NewKey("connection_string", dbConnStr)
+	require.NoError(t, err)
+
+	cfg.HTTPPort = "1300" + strconv.Itoa(instanceId) // 13001 for instance-1, 13002 for instance-2 etc
```

Comment:
It would be safer to use 0, and let system assign the port to avoid conflicts. Same for gRPC and memberlist ports. We will then need to find assigned ports by checking for listener ports.